File: js-python-educative-challenges/two-pointers/sum_of_three_values.py
"""Given an array of integers, nums, and an integer value, target,
    determine if there are any three integers in nums
    whose sum is equal to the target"""


def find_sum_of_three(nums: list[int], target: int) -> bool:
    if len(nums) < 3:
        return False
    sorted_nums = sorted(nums)

    for i in range(len(sorted_nums)-1):
        small_idx = i + 1
        big_idx = len(sorted_nums) - 1

        while small_idx < big_idx:
            res = sorted_nums[i] + sorted_nums[small_idx] + sorted_nums[big_idx]
            if res == target:
                return True
            if res < target:
                small_idx += 1
            if res > target:
                big_idx -= 1
    return False

if __name__ == '__main__':
    nums = [3,7,1,2,8,4,5]
    target = 10
    print(find_sum_of_three(nums, target))

File: js-python-educative-challenges/two-pointers/sum_of_three_values.js
/**
 * Given an array of integers, nums, and an integer value, target, determine
 * if there are any three integers in nums whose sum is equal to the target
 *
 * @param {Array<number>} nums
 * @param {number} target
 * @returns {boolean}
 */

function sumOfThree(nums, target) {
  if (nums.length < 3) return false;
  const sortedNums = nums.slice().sort((a, b) => a > b ? 1 : -1);

  for (i = 0; i < sortedNums.length - 1; i++) {
    let smallIndex = i + 1
    let bigIndex = sortedNums.length - 1
    while (smallIndex < bigIndex) {
      const res = sortedNums[i] + sortedNums[smallIndex] + sortedNums[bigIndex]
      if (res === target) return true;
      if (res < target) smallIndex++;
      if (res > target) bigIndex--;
    }
  }
  return false;
}

// main

const nums = [3,7,1,2,8,4,5]
const target = 10
console.log(sumOfThree(nums, target))
